File: best time to buy and sell stock/sol.py
class Solution(object):
    def maxProfit(self, prices):
        # here the time complexity is O(n^2)
        # max = 0
        # size = len(prices)
        # for i in range(0, size):
        #     temp = prices[i]
        #     for j in range(i, size):
        #         if temp>prices[j]:
        #             continue
        #         else:
        #             if (prices[j] - temp) > max:
        #                 max = prices[j] - temp
        # return max



        buy = prices[0]
        max = 0
        for i in range(1, len(prices)):
            if prices[i] < buy:
                buy = prices[i]
            else:
                temp = prices[i] - buy
                if temp > max:
                    max = temp
        return max

File: best time to buy and sell stock/sol.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int size = prices.size();
        int buy = prices[0];
        int max = 0;

        for(int i=1;i<size;i++)
        {
            if(prices[i] < buy)
            {
                buy = prices[i];
            }
            else
            {
                int temp = prices[i] - buy;
                if(temp > max)
                {
                    max = temp;
                }
            }
        }
        return max;
    }
};
